### PR TITLE
Propagate errors thrown after Asyncify call

### PR DESCRIPTION
**Pull request by RReverser**
Previously any errors thrown after an Asyncify pause would be asynchronously thrown in a global context, with no way to catch and handle them except for `window.onerror`.

This PR builds on top of #14664, catches such errors and reports them as Promise rejections on outermost `ccall` or Embind invocation.

I also replaced `asyncFinalizers` array with a single object, since we forbid using more than one finalize in-progress anyway.

### CODE REVIEW

File: src/library_async.js
Comment (kripken) [inline src/library_async.js:255]: This worries me, as it seems like any exception happening here would be caught and provided as a Promise rejection? For example, if the rewind process itself hits an error. Ideally this would be focused on the actual execution of code.

Or am I looking at this wrong? I admit I have little experience with handling Promise rejections "properly"...

Comment (RReverser) [reply src/library_async.js:255]: > This worries me, as it seems like any exception happening here would be caught and provided as a Promise rejection?

Well yes, but otherwise the error would be thrown into "nowhere" where it can't be caught. We could limit this to errors from the C/C++ code itself, but I'm not sure if it's a better result - even if it's rewind itself that for some reason errored out, it seems useful to throw that error as part of the original call.

Comment (RReverser) [reply src/library_async.js:255]: What are the cases where rewind itself would fail though? I think I've only ever seen errors from unwind (due to OOM of the Asyncify stack size), but never during rewind...

Comment (kripken) [reply src/library_async.js:255]: Yes, the real danger us during unwind, not rewind, that's true. Maybe we don't even have assertions there.

**Review by kripken — APPROVED** [review]
Lgtm if there isn't an easy way to try-catch just on the compiled code. Which seems to be the case.

**Comment by RReverser** [comment]
There are few more issues, converting to a draft for now.

**Comment by RReverser** [comment]
I'm not 100% sure, but... SIMD errors on CI can't be related, can they?

**Comment by kripken** [comment]
SIMD errors were a temporary issue yesterday, please merge in latest `main`.

**Comment by RReverser** [comment]
Ah thanks that fixed it.

**Comment by RReverser** [comment]
Just in case - can you please take another look at changes I've made since (https://github.com/emscripten-core/emscripten/pull/14714/commits/89db9da7b32dc2a0f4794aec15b3bd2780a3638e) before I merge?

Those are to rethrow async errors that didn't happen during exports, but happened during call to e.g. `main`.

File: src/library_async.js
Comment (kripken) [inline src/library_async.js:260]: Can you add a comment for `handled`? Reading the code I'm not sure what that means. (It seems to be `!Asyncify.currData && Asyncify.asyncPromiseHandlers`, but I don't know how to interpret that...)

Comment (RReverser) [reply src/library_async.js:260]: Rather `!Asyncify.currData && !Asyncify.asyncPromiseHandlers` - that is, if there was no handler, then rethrow as regular exception.

Comment (RReverser) [reply src/library_async.js:261]: Added comments.

**Review by kripken — APPROVED** [review]
I see, thanks, lgtm.

**Pull request merged**
